apollo-router/src/plugins/traffic_shaping/mod.rs

```diff
@@ -375,16 +379,23 @@ impl TrafficShaping {
                 tower::retry::RetryLayer::new(retry_policy)
             });
 
+            let timeout = config.shaping.timeout.unwrap_or(DEFAULT_TIMEOUT);
             Either::A(ServiceBuilder::new()
-
                 .option_layer(config.shaping.deduplicate_query.unwrap_or_default().then(
                   QueryDeduplicationLayer::default
                 ))
-                    .layer(TimeoutLayer::new(
-                        config.shaping
-                        .timeout
-                        .unwrap_or(DEFAULT_TIMEOUT),
-                    ))
+                    .map_future_with_request_data(
+                        |req: &subgraph::Request| req.context.clone(),
+                        move |ctx, response| {
+                            request_timeout(timeout, response).unwrap_or_else(|error| {
```

Comment:
ah, I think there was some understanding. When I suggested mapping on the response, that was not to replace the entire `Timeout` layer, but to look at the error it returned and converting it

Reply:
@IvanGoncharov could you bring back the previous timeout code? That was extracted from tower's implementation, we tend to align with their way of doing things in traffic shaping. The only thing we need to do here is converting the error response

Reply:
Yes, I reverted this code.
I still think `map_future_with_request_data` removes the need for extra boilerplate implemented in the tower's layer, but I understand the argument of staying close to the tower's code.